LoopKit/CarbKit/CarbStatus.swift

```diff
@@ -41,32 +41,33 @@ extension CarbStatus: CarbEntry {
 
 
 extension CarbStatus {
-    func dynamicCarbsOnBoard(at date: Date, defaultAbsorptionTime: TimeInterval, delay: TimeInterval, delta: TimeInterval) -> Double {
+    
+    func dynamicCarbsOnBoard(at date: Date, defaultAbsorptionTime: TimeInterval, delay: TimeInterval, delta: TimeInterval, absorptionModel: CarbAbsorptionComputable) -> Double {
         guard date >= startDate - delta,
             let absorption = absorption
         else {
             // We have to have absorption info for dynamic calculation
-            return entry.carbsOnBoard(at: date, defaultAbsorptionTime: defaultAbsorptionTime, delay: delay)
+            return entry.carbsOnBoard(at: date, defaultAbsorptionTime: defaultAbsorptionTime, delay: delay, absorptionModel: absorptionModel)
         }
 
         let unit = HKUnit.gram()
 
-        guard let observedTimeline = observedTimeline else {
-            // Less than minimum observed; calc based on min absorption rate
+        guard let observedTimeline = observedTimeline, let observationEnd = observedTimeline.last?.endDate else {
+            // Less than minimum observed or observation not yet started; calc based on modeled absorption rate
             let total = absorption.total.doubleValue(for: unit)
             let time = date.timeIntervalSince(startDate) - delay
             let absorptionTime = absorption.estimatedDate.duration
-
-            return LinearAbsorption.unabsorbedCarbs(of: total, atTime: time, absorptionTime: absorptionTime)
+            return absorptionModel.unabsorbedCarbs(of: total, atTime: time, absorptionTime: absorptionTime)
         }
 
-        guard let end = observedTimeline.last?.endDate, date <= end else {
+        guard date <= observationEnd else {
             // Predicted absorption for remaining carbs, post-observation
-            let total = absorption.remaining.doubleValue(for: unit)
-            let time = date.timeIntervalSince(absorption.observedDate.end)
-            let absorptionTime = absorption.estimatedTimeRemaining
-
-            return LinearAbsorption.unabsorbedCarbs(of: total, atTime: time, absorptionTime: absorptionTime)
+            let effectiveTime = date.timeIntervalSince(observationEnd) + absorption.timeToAbsorbObservedCarbs
```

Comment:
All "time" is a time interval (unix time = seconds since unix epoch, time of day = time since midnight, etc). This is a pattern you'll see elsewhere in Loop; a TimeInterval used as a "time", like time of day, or in other contexts where there is a strong assumed reference point.  All times are measurements of amount of time, relative to some reference point.